[MRG + 1] Fix perplexity method by adding _unnormalized_transform method, Issue #7954

garyForeman opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
Fixes #7954 

#### What does this implement/fix? Explain your changes.
This fixes the broken perplexity method of the LatentDirichletAllocation class. This method was broken in the 0.18 release when the default behavior of the transform method switched to returning normalized document topic distributions. However, the perplexity calculation uses likelihoods rather than probabilities.

To fix the issue, I have added an optional argument, normalize, to the transform method. The default is set to True such that if the argument is not specified, the behavior of the transform method remains unchanged from the version 0.18 release. When the transform method is called from the perplexity method, normalize=False is passed so that the likelihoods rather than the probabilities are returned.

#### Any other comments?
While I believe the fix proposed here is the most elegant, I would understand if we didn't want to add an additional argument to the transform method. If this is the case, I am happy to pivot the implementation to use a "private" class attribute that stores the normalization information.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Fix perplexity method by adding normalize argument to transform method

amueller commented:
Thank you for the fix.

Can you check on the failing tests?
We also need a regression test for this bug.

I think we don't want to add to the public api for this (imho), and I don't think we want to store the state in a private variable.

I think my preferred solution would be to add a private method ``_transform`` or ``_unnormalized_transform`` that is called both in ``transform`` and ``perplexity``.

garyForeman commented:
@amueller I really like your solution, which is not something I would have come up with on my own. I'll work on implementing your suggestion. Thanks for the advice!

FYI, the failing tests are to do with the fact that I've changed the api, which, as you've suggested, is not ideal. Specifically, the test failed when the fit_transform method implemented in TransformMixin had no way to pass the normalize argument when it called transform.

amueller commented:
Ok lets see how it goes when you implement the suggestion :)

garyForeman commented:
I've implemented an `_unnormalized_transform` method as you've suggested, which technically passes the test module and fixes the `perplexity` method when passed `doc_topic_distr=None`. However, if a user saves the output of `transform` and passes the result to `perplexity`, the result will be incorrect because `transform` will only provide a *normalized* document topic distribution. How would you like to handle this situation?

amueller commented:
That's a good question and I don't really have a straight-forward answer.
I find the API for ``perplexity`` with the optional ``doc_topic_distr`` a bit odd, because we have no way of checking whether that was computed correctly.
Also, that name seems a bit confusing because it's not the normalized distribution - though I'm not that familiar with the language of LDA and might get it wrong.

We could deprecate passing the precomputed likelihood in the public interface. It might have been that it was only added there for the use in the training algorithm, but I'm not sure.

Alternatively we could add a check that throws a warning if someone passes normalized probabilities to ``perplexity``, though that seems to be a bit of a hack to me.

garyForeman commented:
I'm inclined to agree with the deprecation of the optional `doc_topic_distr` argument especially considering that as of sklearn v0.18, the user no longer has any way to access an appropriately unnormalized distribution. Would you like me to resubmit what I currently have written?

Removing the argument entirely would require some edits to the testing module. If this is the route we want to go, I'll look into which lines of `test_online_lda.py` would be affected by the removal of the `doc_topic_distr` argument.

amueller commented:
>  Would you like me to resubmit what I currently have written?

Not sure what you mean. You can update the branch with the appropriate changes.
You can go ahead and make the changes but I'd like someone else to chime in if they think that's the best solution. @jnothman ?

Commit:
Add _unnormalized_transform method that returns an unnormalized document topic distribution

Commit:
Update LatentDirichletAllocation score method to use _unnormalized_transform

Commit:
Update test_lda_score_perplexity and test_perplexity_intput_format to use _unnormalized_transform to set distr

Commit:
Remove superfluous white space

amueller commented:
So that is ok as a fix, but now the API is a bit weird as you agreed, so I think you should go ahead and deprecate ``doc_topic_distr``

garyForeman commented:
Sounds good, I'll get to it in the next day or so.

Commit:
Depricate doc_topic_distr argument in perplexity method

Commit:
Modify testing module to allow for doc_topic_distr deprication

garyForeman commented:
Failed test was sklearn/decomposition/test_pca.py in Python 3 on 32-bit architecture. My submission was accurate to 11 decimal places, but accuracy required is 12. This seems a bit harsh considering 32-bit machine epsilon is about 5.96E-8 http://bit.ly/2hvHU6f

Commit:
Merge branch 'master' into perplexity_fix

jnothman commented:
Should this now be MRG?

garyForeman commented:
I wasn't planning to make any more updates unless you or @amueller had further suggestions. So in my biased opinion, it's ready to merge :)

garyForeman commented on sklearn/decomposition/online_lda.py:
```diff
@@ -719,3 +739,21 @@ def perplexity(self, X, doc_topic_distr=None, sub_sampling=False):
 
         return np.exp(-1.0 * perword_bound)
+
+    def perplexity(self, X, sub_sampling=False):
```
Ok, makes sense. Sorry for the confusion.

Commit:
Reintroduce doc_topic_distr argument but throw a DeprecationWarning if it is not set to None

Commit:
Merge branch 'master' into perplexity_fix

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -719,3 +739,30 @@ def perplexity(self, X, doc_topic_distr=None, sub_sampling=False):
+        """
+
+        if doc_topic_distr is not None:
+            DeprecationWarning("Argument 'doc_topic_distr' is deprecated as "
```
Maybe say explicitly that it is ignored.

amueller commented:
Please add a regression test that checks that the perplexity is now computed correctly. I think the best way would be to check against the one computed during fit, but I'm not sure.

Also, please add an entry to whatsnew, and add a ``versionchanged`` entry to ``transform`` that mentions the deprecation of the argument.

Commit:
Remove superfluous import

Commit:
Document perplexity bug fix as well as doc_topic_distr deprecation

Commit:
Add regression test to ensure consistency between perplexity calculation made during the fit method and the value returned by the perplexity method itself.

Commit:
Remove uneeded batch_size variable from test_lda_fit_perplexity

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -295,26 +295,46 @@ def test_perplexity_input_format():
+    # returned by the perplexity method
+    n_topics, X = _build_sparse_mtx()
+    lda = LatentDirichletAllocation(n_topics=n_topics, max_iter=1,
+                                     learning_method='batch', random_state=0,
```
nitpick: this line and the next seem to be indented one space too much.

amueller commented on sklearn/decomposition/tests/test_online_lda.py:
```diff
@@ -295,26 +295,46 @@ def test_perplexity_input_format():
+                                     evaluate_every=1)
+                                    
+    # The following is taken from the body of the fit method
+    X = lda._check_non_neg_array(X, "LatentDirichletAllocation.fit")
```
hm this is not a very robust way to test this, but I also don't really see a better way.
We could store the ``bound`` variable in ``fit`` and compare against that?
``self.bound_ = bound``?

In general I do like storing the objective. Though we might not have calculated it in the last iteration... hm...

Commit:
Remove extra whitespace

Commit:
Add calculation and storage of final perplexity value on the training set in fit method

Commit:
Update test_lda_fit_perplexity to use lda.bound_ attribute to refer to perplexity found while running fit method

Commit:
Fix lines that were >79 characters

garyForeman commented:
Added a call to `_e_step` and `_perplexity_precomp_distr` just above `return self` in the `fit` method. This calculates the perplexity on the training data after fitting is completed and stores to `self.bound_` as you suggested. Modified `test_lda_fit_perplexity` to use this added functionality.

amueller commented on sklearn/decomposition/online_lda.py:
```diff
@@ -541,10 +541,18 @@ def fit(self, X, y=None):
                 self.n_iter_ += 1
+
+        # calculate final perplexity value on train set
+        doc_topics_distr, _ = self._e_step(X, cal_sstats=False,
```
I mean it adds a little overhead, but I think that's fine?

amueller commented:
LGTM

garyForeman commented:
Great! Thanks @amueller for your patience and you help with my first scikit-learn contribution.

amueller commented:
@garyForeman thanks for your contribution :) and for your patience with our slow process....

jnothman commented:
I'm not very familiar with this part of the codebase, but will review soon

amueller commented:
@jnothman no-one is, i think, and it's messy :-/ we should try to simplify but that's for later.

jnothman commented on sklearn/decomposition/online_lda.py:
```diff
@@ -719,3 +750,34 @@ def perplexity(self, X, doc_topic_distr=None, sub_sampling=False):
+        """
+
+        if doc_topic_distr is not None:
+            warnings.warn("Argument 'doc_topic_distr' is deprecated and will "
```
Technically this should also show if the user provides an explicit value for `doc_topic_distr`. So you need a sentinel other than `None`, such as `'deprecated'`

jnothman replied:
yes

jnothman reviewed (commented):
Otherwise LGTM, with the caveat that I'm not entirely confident that this is the right fix, as I don't know this part of the code.

Commit:
Merge branch 'master' into perplexity_fix

Commit:
Move description of doc_topic_distr deprecation to API changes summary

jnothman commented on sklearn/decomposition/online_lda.py:
```diff
@@ -572,8 +580,29 @@ def transform(self, X):
+        doc_topic_distr : shape=(n_samples, n_topics)
+            Document topic distribution for X.
+        """
+
```
I feel like there's more whitespace than makes this readable, when there's an extra line between function docstring's closing `"""` on a line of its own and function code. Also blank line between calculating a value and returning it seems unhelpful for readability.

amueller replied:
uh sorry I misread you mean "this blank line and the next blank line". Never mind.

jnothman commented:
It's alright. As wasted as the words may seem, I don't mind justifying
what style I find elegant.

On 15 December 2016 at 07:30, Andreas Mueller <notifications@github.com>
wrote:

> *@amueller* commented on this pull request.
> ------------------------------
>
> In sklearn/decomposition/online_lda.py
> <https://github.com/scikit-learn/scikit-learn/pull/7992>:
>
> > +        """Transform data X according to the fitted model.
> +
> +           .. versionchanged:: 0.18
> +              *doc_topic_distr* is now normalized
> +
> +        Parameters
> +        ----------
> +        X : array-like or sparse matrix, shape=(n_samples, n_features)
> +            Document word matrix.
> +
> +        Returns
> +        -------
> +        doc_topic_distr : shape=(n_samples, n_topics)
> +            Document topic distribution for X.
> +        """
> +
>
> uh sorry I misread you mean "this blank line and the next blank line".
> Never mind.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7992>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63FEkuiTIDjJ3bkDKyuZ-psCJIgUks5rIFHigaJpZM4LFiE9>
> .
>

Commit:
Change doc_topic_distr default to 'deprecated' in perplexity method

jnothman commented on sklearn/decomposition/online_lda.py:
```diff
@@ -719,3 +746,33 @@ def perplexity(self, X, doc_topic_distr=None, sub_sampling=False):
+        doc_topic_distr : None or array, shape=(n_samples, n_topics)
+            Document topic distribution.
+            If it is None, it will be generated by applying transform on X.
+
```
you can actually use a `.. deprecated` note here

jnothman commented:
@amueller, we're not going to get someone more familiar with this code to comment, are we?

Commit:
Add .. debrecated note in perplexity doc string under doc_topic_distr description

amueller commented:
we could try to summon @ogrisel or @larsmans but I'm not even sure how familiar they are with the code. @chyikwei is the original author.

chyikwei commented:
The fix looks good to me.

We need to pass unnormalized `doc_topic_distr` to estimate the lower bound of perplexity, and that's why `transform` method return unnormalized value before. (Eq.16  in the [reference paper](https://www.cs.princeton.edu/~blei/papers/HoffmanBleiBach2010b.pdf)).

For `perplexity` function, it makes sense to deprecate `doc_topic_distr` since `transform` doesn't return it anymore. And  maybe we should also deprecate `sub_sampling`. (It is only used in online method)

jnothman commented:
Ideally, we'd also check that the warning is raised when `doc_topic_distr` is passed (None or otherwise). Otherwise, we should merge this ASAP

garyForeman commented:
I'll go ahead and write a regression test for the deprecated `doc_topic_distr` warning message.

IHMO, deprecating the `sub_sampling` argument, sounds like an issue for a separate pull request. After a quick read through the code, it looked as though `sub_sampling` is set to `False` everywhere, but I wouldn't say I have a great understanding of the implications behind deprecating this argument.

chyikwei commented:
yeah. deprecate `sub_sampling` should be a separate issue.

Commit:
Add regression test for doc_topic_distr deprecation warning message

Commit:
Merge branch 'master' into perplexity_fix

jnothman commented:
Thanks a lot @garyForeman, and @chyikwei, your continued advice is much appreciated.

naoyak commented:
@amueller @garyForeman 

I was discussing with @jnothman a similar PR (https://github.com/scikit-learn/scikit-learn/pull/8137#issuecomment-269897136) and it came up that this change doesn't really follow the standard deprecation procedure.

I think I understand the justification that the parameter is pretty much useless if the user isn't able to obtain and supply a valid unnormalized distribution, but maybe the functionality should be restored, or at least a __loud and clear__ notice in the docs and the `DeprecationWarning` that passing `doc_topic_distr` does *nothing*?

Sorry to butt in on a topic I know little about and just wanted to relay that input.

garyForeman commented:
@naoyak @jnothman 

I'm not sure I would say that passing `doc_topic_distr` does *nothing*. The code executes as if the argument was not passed by computing the properly *unnormalized* topic distribution, which is needed in order to produce correct results.

#8137 appears to be focused on tidying up the API, whereas this is a bug fix. It makes sense to maintain the `ridge_alpha` parameter in the near future, otherwise, you will break users' code. This change does not break users' code, but will admittedly increase run time.

As @amueller and I discussed above, the ideas we had for maintaining working functionality of the `doc_topic_distr` were to either change the API or store normalization constants in a private attribute. Neither idea was palatable.

jnothman commented:
Sorry for not reviewing carefully. If doc_topic_distr is not being passed on to `_perplexity_precomp_distr`, the warning should say that it is *already* being ignored.

garyForeman commented:
Ok, I can add that in. I'll open a new report request shortly.

garyForeman commented on sklearn/decomposition/online_lda.py:
```diff
@@ -719,3 +746,35 @@ def perplexity(self, X, doc_topic_distr=None, sub_sampling=False):
+        """
+        if doc_topic_distr != 'deprecated':
+            warnings.warn("Argument 'doc_topic_distr' is deprecated and will "
+                          "be ignored as of 0.19. Support for this argument "
```
So I'll change this from future to present tense, i.e. "...is deprecated and is being ignored as of 0.19." Does that address the issue?

jnothman commented:
If there's no way to keep using doc_topic_distr then yes.

On 3 January 2017 at 09:14, Gary Foreman <notifications@github.com> wrote:

> *@garyForeman* commented on this pull request.
> ------------------------------
>
> In sklearn/decomposition/online_lda.py
> <https://github.com/scikit-learn/scikit-learn/pull/7992#pullrequestreview-14874217>
> :
>
> > +            Document word matrix.
> +
> +        doc_topic_distr : None or array, shape=(n_samples, n_topics)
> +            Document topic distribution.
> +            If it is None, it will be generated by applying transform on X.
> +
> +            .. deprecated:: 0.19
> +
> +        Returns
> +        -------
> +        score : float
> +            Perplexity score.
> +        """
> +        if doc_topic_distr != 'deprecated':
> +            warnings.warn("Argument 'doc_topic_distr' is deprecated and will "
> +                          "be ignored as of 0.19. Support for this argument "
>
> So I'll change this from future to present tense, i.e. "...is deprecated
> and is being ignored as of 0.19." Does that address the issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7992#pullrequestreview-14874217>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6yOw8k4qtQNKVxhz1kjrX0wtMkmZks5rOXa0gaJpZM4LFiE9>
> .
>

naoyak commented on sklearn/decomposition/online_lda.py:
```diff
@@ -719,3 +746,35 @@ def perplexity(self, X, doc_topic_distr=None, sub_sampling=False):
+        X : array-like or sparse matrix, [n_samples, n_features]
+            Document word matrix.
+
+        doc_topic_distr : None or array, shape=(n_samples, n_topics)
```
The docstring should probably note clearly that `doc_topic_distr` is currently discarded as well, since that's where users look first.

jnothman commented:
well it doesn't do anything, the docstring should certainly not say that it
does

On 3 Jan 2017 9:57 am, "Naoya Kanai" <notifications@github.com> wrote:

> *@naoyak* commented on this pull request.
> ------------------------------
>
> In sklearn/decomposition/online_lda.py
> <https://github.com/scikit-learn/scikit-learn/pull/7992#pullrequestreview-14875537>
> :
>
> > +
> +    def perplexity(self, X, doc_topic_distr='deprecated', sub_sampling=False):
> +        """Calculate approximate perplexity for data X.
> +
> +        Perplexity is defined as exp(-1. * log-likelihood per word)
> +
> +        .. versionchanged:: 0.19
> +           *doc_topic_distr* argument has been depricated because user no
> +           longer has access to unnormalized distribution
> +
> +        Parameters
> +        ----------
> +        X : array-like or sparse matrix, [n_samples, n_features]
> +            Document word matrix.
> +
> +        doc_topic_distr : None or array, shape=(n_samples, n_topics)
>
> The docstring should probably note clearly that doc_topic_distr is
> currently discarded as well, since that's where users look first.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7992#pullrequestreview-14875537>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_dgrRQ2e47ghPPdgo1B8pWztjHyks5rOYDNgaJpZM4LFiE9>
> .
>
